content/blog/pulumi-convert-terraform-improvements/index.md

```diff
@@ -0,0 +1,351 @@
+
+```diff
+metadata={
+-        "DB_HOST": db["url"],
++        "DB_HOST": db.url,
+        "DB_USER": "root",
+-        "DB_PASS": db_password["plaintext"],
+-        "DB_NAME": db["name"],
++        "DB_PASS": db_password.plaintext,
++        "DB_NAME": db.name,
+}
+```
+
+{{% /choosable %}}
+
+{{% choosable language go %}}
+
+```shell
+pulumi convert --from terraform --language go --out pulumi-go-program
+```
+
+I have opted to output it to a different directory to preserve the contents of my Terraform project.
+
+### Project structure
+
+If we navigate to the `golang` directory, we can see a few things:
+
+- main.go, our equivalent to main.tf for golang projects
+- sdks directory, where the generated bridged provider is output
+- other golang project artifacts (go.mod, go.sum)
+
+### Go specific cleanup
+
+The go code generator outputs everything in the Terraform code, even if it is
+unused. This is an error in a go program, so I had to manually remove
+unreferenced variables like GCP project, region, and PlanetScale service token.
+
+These are part of the output because in it's current iteration the code
+converter will convert everything, even if ultimately it is provider
+configuration (see [cleanup](#cleanup)) and not actual code.
+
+{{% /choosable %}}
+
+{{% /chooser %}}
+
+### Cleanup
+
+If we simply run `pulumi preview` this project unfortunately won't run right away.  
+We need to create our stack configuration that contains the GCP and PlanetScale credentials.
+
+These were part of our `main.tf` file, but the code generator doesn't
+distinguish between provider config and code, so we need to handle this
+part on our own (for now).
+
+These can go in either your `Pulumi.yaml` or your stack configuration.  Places
+for them are already generated, you just need to fill them in:
+
+```yaml
+name: terraform-convert-example
+runtime: nodejs
+config:
+  google:project:
+    value: 'TODO: var.gcp_project' # fill in here
+  google:region:
+    value: 'TODO: var.gcp_region'
+  google:zone:
+    value: 'TODO: var.gcp_zone'
+  planetscale:serviceToken:
+    value: 'TODO: var.planetscale_service_token'
+  planetscale:serviceTokenName:
+    value: planetscaletoken
+```
+
+Also notice the configuration names differ from the pulumi counterparts "google" should be "gcp".
+
+After filling things in the `Pulumi.yaml` should look something like this:
+
+```yaml
+name: terraform-convert-example
+runtime: nodejs
+config:
+  gcp:project:
+    value: 'brandonpollack23'
+  gcp:region:
+    value: 'us-central1'
+  gcp:zone:
+    value: 'us-central1-a'
+  planetscale:serviceToken:
+    value: 'supersecrettoken'
+  planetscale:serviceTokenName:
+    value: planetscaletoken
+```
+
+and if you run `pulumi preview` it successfully generates a plan!
+
+There is still more you can do, the generated code could be cleaned up some as there
+are some unused variables, etc.
+
+## Limitations
```

Comment:
We wanted to also mention `pulumi import` support for any Terraform/OpenTofu provider in this blog post as well. I think that could be as simple as an extra section above that says something like:

> ## Importing
>
> We also now support importing existing infrastructure from any Terraform/OpenTofu provider with the `pulumi import` command. For example:
> (insert some example `pulumi import` command)

It doesn't necessarily have to go as in depth as the sections on `pulumi convert` above.

Let me know if you want to take a crack at writing this or if you'd like some help.